python/feathub/registries/mysql_registry.py

```diff
@@ -0,0 +1,273 @@
+
+PASSWORD_CONFIG = MYSQL_REGISTRY_PREFIX + "password"
+PASSWORD_DOC = "The password of the user."
+
+
+mysql_registry_config_defs: List[ConfigDef] = [
+    ConfigDef(
+        name=DATABASE_CONFIG,
+        value_type=str,
+        description=DATABASE_DOC,
+        default_value=None,
+    ),
+    ConfigDef(
+        name=TABLE_CONFIG,
+        value_type=str,
+        description=TABLE_DOC,
+        default_value=None,
+    ),
+    ConfigDef(
+        name=HOST_CONFIG,
+        value_type=str,
+        description=HOST_DOC,
+        default_value=None,
+    ),
+    ConfigDef(
+        name=PORT_CONFIG,
+        value_type=int,
+        description=PORT_DOC,
+        default_value=3306,
+    ),
+    ConfigDef(
+        name=USERNAME_CONFIG,
+        value_type=str,
+        description=USERNAME_DOC,
+        default_value=None,
+    ),
+    ConfigDef(
+        name=PASSWORD_CONFIG,
+        value_type=str,
+        description=PASSWORD_DOC,
+        default_value=None,
+    ),
+]
+
+
+class MySqlRegistryConfig(RegistryConfig):
+    def __init__(self, props: Dict[str, Any]) -> None:
+        super().__init__(props)
+        self.update_config_values(mysql_registry_config_defs)
+
+
+def _get_digest(json_string: str) -> str:
+    return sha256(json_string.encode("utf-8")).hexdigest()
+
+
+class MySqlRegistry(Registry):
+    """
+    A registry that stores entities in a MySQL database.
+    """
+
+    def __init__(
+        self,
+        props: Dict,
+    ):
+        super().__init__("mysql", props)
+        mysql_registry_config = MySqlRegistryConfig(props)
+        self.database = mysql_registry_config.get(DATABASE_CONFIG)
+        self.table = mysql_registry_config.get(TABLE_CONFIG)
+        self.host = mysql_registry_config.get(HOST_CONFIG)
+        self.port = mysql_registry_config.get(PORT_CONFIG)
+        self.username = mysql_registry_config.get(USERNAME_CONFIG)
+        self.password = mysql_registry_config.get(PASSWORD_CONFIG)
+
+        # dict that acts as a local cache for built and registered tables.
+        # each value in the dict is a tuple of table before build and table after build.
+        self.tables: Dict[str, Tuple[TableDescriptor, TableDescriptor]] = {}
+
+        self.conn = mysql.connector.connect(
+            host=self.host,
+            port=self.port,
+            user=self.username,
+            password=self.password,
+            database=self.database,
+        )
+        self.cursor = self.conn.cursor()
+
+        self.cursor.execute(
+            f"""
+                CREATE TABLE IF NOT EXISTS `{self.table}`(
+                   `digest` VARCHAR(64) NOT NULL,
+                   `name` TEXT NOT NULL,
+                   `timestamp` TIMESTAMP NOT NULL,
+                   `is_deleted` BOOLEAN NOT NULL,
+                   `original_json` TEXT NOT NULL,
+                   `built_json` TEXT NOT NULL,
+                   PRIMARY KEY ( `digest`, `timestamp` )
```

Comment:
How about using `name` and `timestamp` as the primary keys?

Reply:
MySQL requires that columns that are used as keys must have a limited length, while Feathub does not have this requirement for `name` strings.

Reply:
MySQL requires that key columns have a fixed length, while Feathub has not yet limited the size of feature descriptor names. I have required that descriptors to be registered into MySQL must have their names less than 64 characters, and added a TODO that we may apply this requirement to all Features and TableDescriptors in the future.

Reply:
Sounds good.